Analysis and applications of partitioning (#24)
intarchboard/draft-obliviousness: draft-kuepau-iab-privacy-partitioning.md

# Applying Privacy Partioning

> Applying privacy partitioning to an existing or new system or protocol is a straightforward process.

Author: Maybe we say it's a conceptually simple process? As in, "easier said than done"?

> Applying privacy partitioning to an existing or new system or protocol is a straightforward process.

Reviewer: ```suggestion
Applying privacy partitioning to an existing or new system or protocol requires the following steps:
```

> 1. Identify all types of information used or exposed in a system or protocol, some of which may be user-identifying.

Reviewer: ```suggestion
1. Identify the types of information used or exposed in a system or protocol, some of which can be used to identity a user or correlate to other contexts.
```

> 1. Identify the types of information used or exposed in a system or protocol, some of which can be used to identity a user or correlate to other contexts.

Reviewer: ```suggestion
1. Identify the types of information used or exposed in a system or protocol, some of which can be used to identify a user or correlate to other contexts.
```

> 1. Partition different types of user-identifying information across contexts to minimize the amount of

Reviewer: I'm concerned that just partitioning out the data that is obviously user-identifying isn't enough; how can we capture the correlation angle too?

Author: I think your suggested text probably covers it, which states that two contexts must not be trivially correlated through any shared or common info.

> 1. Partition different types of user-identifying information across contexts to minimize the amount of
>    user-identifying information in any single context.
> 

Reviewer: ```suggestion
1. Partition different types of user-identifying or correlatable information across contexts to minimize the amount of
user-identifying or correlatable information in any single context.

```

> user-identifying or correlatable information in any single context.

Reviewer 2: > minimize the amount of  user-identifying or correlatable information in any single context.
Per discussion slack, I actually think this is not fully correct and also not what we say earlier. Earlier we talk about separate the who from the what and minimising amount of information needed to fulfill a function to only reveal user-specific information to parties that need access to the information for their intended purpose.

Author: (Is this now resolved?)

> 1. Partition different types of user-identifying or correlatable information across contexts to minimize the amount of
> user-identifying or correlatable information in any single context.

Reviewer 2: ```suggestion
1.  Separate user-identifying information from other user data in different contexts and reduce the amount of
user-identifying or correlatable information in each context to the minimum needed to fulfil the intended function.
```

Author: I think this is now resolved based on the change below?

Reviewer 2: No, this is a separate point: "minimising amount of user-identifying or correlatable information in any single context" is a too strong goal. Earlier we say you want to "minimise to the information needed to fulfil a certain function". That's something different.

> 1. Partition different types of user-identifying or correlatable information across contexts to minimize the amount of
> user-identifying or correlatable information in any single context.

Reviewer 2: ```suggestion
1.  Separate user-identifying information from other user data in different contexts and reduce the amount of
user-identifying or correlatable information in each context to the minimum needed to fulfil the intended function.
```

> The most impactful types of user-identifying information to partition are (a) user identity and (b) user data.

Reviewer: ```suggestion
The most impactful types of user-identifying information to partition are (a) user identity or identities that can be linked and (b) user data.
```

> The most impactful types of user-identifying information to partition are (a) user identity or identities that can be linked and (b) user data.

Author: I'm not sure I follow you, since that's exactly what the text here says. That is, it says the most important things to separate are information relating to user identity (IP address) and user data (HTTP request).

> The most impactful types of user-identifying information to partition are (a) user identity or identities that can be linked and (b) user data.

Reviewer 2: ```suggestion
The most impactful partitioning is separating user-identifying information from other user data.
```

> The most impactful types of user-identifying information to partition are (a) user identity or identities that can be linked and (b) user data.

Reviewer 2: it's not quite what the text says because it currently says that user data (b) is also user-identifying information which is confusing.

> The most impactful types of user-identifying information to partition are (a) user identity or identities that can be linked and (b) user data.

Author: User data _can_ be user-identifying, since it's data about or generated by the user. This is an important point I don't think we can omit. How about the following?

```
The most impactful types of information to partition are (a) user identity or identities that can be linked and (b) user data, which can be user-identifying for certain applications.
```

> The most impactful types of user-identifying information to partition are (a) user identity or identities that can be linked and (b) user data.

Reviewer 2: I guess this sentence is not wrong but not sure it's the point. E.g. if I visit videosofcats.com with my IP address, someone who observes the traffic knows that I love cats. So separating this information is useful. If I visit myprivatepageofcatvideos.org where I'm the only user ever visiting this page, an observer can identify me no matter if they see my IP address or not. Hiding the IP address might still be important to not correlate this traffic with other traffic. But isn't the first case the most/more important one, where the user data in itself in not user-identifying?

> The most impactful types of user-identifying information to partition are (a) user identity or identities that can be linked and (b) user data.

Author: Yep, and that's an example of how this principle, even if applied correctly, is insufficient in many cases. This PR does acknowledge this limitation, so I don't think further changes are needed here.

> The most impactful types of user-identifying information to partition are (a) user identity or identities that can be linked and (b) user data.

Reviewer 2: So why is “which can be user-identifying for certain applications” needed?

> The most impactful types of user-identifying information to partition are (a) user identity or identities that can be linked and (b) user data.

Reviewer 2: isn't the most impactful partition to separate user-identifying data from non-user-identifying data?

> The most impactful types of user-identifying information to partition are (a) user identity or identities that can be linked and (b) user data.

Author: ```suggestion
The most impactful types of information to partition are (a) user identity or identities (such as an account name or IP address) that can be linked and (b) user data (such as the content a user is accessing), which can be often sensitive when combined with user identity. Note that user data can itself be user-identifying, in which case it should be treated as an identifier.
```

For example, Oblivious DoH and Oblivious HTTP partition the client IP address and client request data into
separate contexts, thereby ensuring that no entity beyond the client can observe both. Collusing across contexts
may reverses this partition process, but can also promote non-user-identifying information to user-identifying.
For example, in CONNECT proxy systems that use QUIC, the QUIC connection ID is inherently non-user-identifying
since it is generated randomly {{QUIC, Section 5.1}}. However, if combined with another context that has user-identifying
information such as the client IP address, the QUIC connection ID can become user-identifying information.

> This partitioning process is inherently imperfect, and can be applied incorrectly. Partitions may contain

Reviewer: ```suggestion
This partitioning process can be applied incorrectly or incompletely. Partitions may contain
```

> This partitioning process can be applied incorrectly or incompletely. Partitions may contain

Author: Yeah, but I'll change this to "context" for now. We can converge on a standard term later.

more user-identifying information than desired, or some information in a context may be more user-identifying

> than intended. Nevertheless, partitions can help improve the client's privacy posture when applied correctly.

Reviewer: ```suggestion
than intended. Nevertheless, partitions can help improve the client's privacy posture when applied carefully.
```

> This partitioning process can be applied incorrectly or incompletely. Partitions may contain

Author: ```suggestion
This partitioning process can be applied incorrectly or incompletely. Contexts may contain
```

Evaluating and qualifying the resulting privacy of a system or protocol that applies privacy partitioning depends

> on the contexts that exist and types of user-identifying information in each context. Such evaluation is

Reviewer 2: I proposing this change because it is not at all clear what is means to evaluate the context that exists. I would say this evaluation is actually to identify the contexts itself, which means you identify all entities that below to a context and the information that is shared in this context.

# Applying Privacy Partioning

Applying privacy partitioning to an existing or new system or protocol requires the following steps:

1. Identify the types of information used or exposed in a system or protocol, some of which can be used to identify a user or correlate to other contexts.

> 1.  Separate user-identifying information from other user data in different contexts and reduce the amount of
> user-identifying or correlatable information in each context to the minimum needed to fulfil the intended function.

Reviewer 2: ```suggestion
1. Partition user data such that the amount of user-identifying or correlatable information in each context is reduced to the minimum needed to fulfil the intended function.
```

> 1.  Separate user-identifying information from other user data in different contexts and reduce the amount of
> user-identifying or correlatable information in each context to the minimum needed to fulfil the intended function.

Author: ```suggestion
1.  Partition data to minimize the amount of user-identifying or correlatable information in any given context to only include what is necessary for that context, and prevent sharing of data across contexts wherever possible.
```
